src/keys.rs

```diff
@@ -1,70 +1,193 @@
+    /// Returns `None` if the bytes do not correspond to a valid Orchard spending key.
+    pub fn from_bytes(sk: [u8; 32]) -> CtOption<Self> {
+        let sk = SpendingKey(sk);
+        // If ask = 0, discard this key.
+        let ask = SpendAuthorizingKey::derive_inner(&sk);
+        CtOption::new(sk, !ask.ct_is_zero())
+    }
+}
 
+/// A spending authorizing key, used to create spend authorization signatures.
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+///
+/// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
 #[derive(Debug)]
-pub(crate) struct SpendAuthorizingKey;
+pub(crate) struct SpendAuthorizingKey(redpallas::SigningKey<SpendAuth>);
+
+impl SpendAuthorizingKey {
+    /// Derives ask from sk. Internal use only, does not enforce all constraints.
+    fn derive_inner(sk: &SpendingKey) -> pallas::Scalar {
+        to_scalar(prf_expand(&sk.0, &[0x06]))
+    }
+}
 
 impl From<&SpendingKey> for SpendAuthorizingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        let ask = Self::derive_inner(sk);
+        // SpendingKey cannot be constructed such that this assertion would fail.
+        assert!(!bool::from(ask.ct_is_zero()));
+        // TODO: Add TryFrom<S::Scalar> for SpendAuthorizingKey.
+        let ret = SpendAuthorizingKey(ask.to_bytes().try_into().unwrap());
+        // If the last bit of repr_P(ak) is 1, negate ask.
+        if (<[u8; 32]>::from(AuthorizingKey::from(&ret).0)[31] >> 7) == 1 {
+            SpendAuthorizingKey((-ask).to_bytes().try_into().unwrap())
+        } else {
+            ret
+        }
     }
 }
 
 /// TODO: This is its protocol spec name for Sapling, but I'd prefer a different name.
 #[derive(Debug)]
-pub(crate) struct AuthorizingKey;
+pub(crate) struct AuthorizingKey(redpallas::VerificationKey<SpendAuth>);
 
 impl From<&SpendAuthorizingKey> for AuthorizingKey {
-    fn from(_: &SpendAuthorizingKey) -> Self {
-        todo!()
+    fn from(ask: &SpendAuthorizingKey) -> Self {
+        AuthorizingKey((&ask.0).into())
     }
 }
 
+/// A key used to derive [`Nullifier`]s from [`Note`]s.
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+///
+/// [`Nullifier`]: crate::note::Nullifier;
+/// [`Note`]: crate::note::Note;
+/// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
 #[derive(Debug)]
-pub(crate) struct NullifierDerivingKey;
+pub(crate) struct NullifierDerivingKey(pallas::Base);
 
 impl From<&SpendingKey> for NullifierDerivingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        NullifierDerivingKey(to_base(prf_expand(&sk.0, &[0x07])))
     }
 }
 
 /// A key that provides the capability to view incoming and outgoing transactions.
 ///
 /// This key is useful anywhere you need to maintain accurate balance, but do not want the
 /// ability to spend funds (such as a view-only wallet).
```

Comment:
```suggestion
/// An Orchard full viewing key, as defined in [Zcash Protocol Spec §4.2.3: Orchard Full Viewing Keys][orchardfullviewingkeyencoding]
///
/// Provides the capability to view incoming and outgoing transactions. This key is 
/// useful anywhere you need to maintain accurate balance, but do not want the ability 
/// to spend funds (such as a view-only wallet).
///
/// [orchardfullviewingkeyencoding]: https://zips.z.cash/protocol/nu5.pdf#orchardfullviewingkeyencoding
```

Is there a better section in the spec? Also the line wrap might be wonky

Reply:
That section of the specification defines the string encoding of `fvk`, not `fvk` itself.

I'm not sure yet whether it makes sense to implement string encoding in this crate or the downstream users (`zcash_primitives` and `zebra-chain`), since doing so requires knowledge of the network type, which the downstream crates define but the Orchard protocol is agnostic over. It would be nice if the two downstream crates could agree on a network type, but that seems like it might require a larger refactor beyond Orchard.

Reply:
> That section of the specification defines the string encoding of `fvk`, not `fvk` itself.
> 
> I'm not sure yet whether it makes sense to implement string encoding in this crate or the downstream users (`zcash_primitives` and `zebra-chain`), since doing so requires knowledge of the network type, which the downstream crates define but the Orchard protocol is agnostic over. It would be nice if the two downstream crates could agree on a network type, but that seems like it might require a larger refactor beyond Orchard.

Oh very agree. I think there is no better layout/description of the Orchard Full Viewing Key than this section at present, the best other section i can see in version v2021.1.16-10-g928b3d is https://zips.z.cash/protocol/nu5.pdf#addressesandkeys, unless I'm missing something or there are incoming additions. I think this is a holdover from Sapling but.

Reply:
oof that may mean we can't just impl FromStr and Display on this, we'd have to wrap it or transform from this key: https://doc.zebra.zfnd.org/src/zebra_chain/sapling/keys.rs.html#684

Reply:
Use https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents for the spec link.

Reply:
I started putting together a `zcash_address` crate which I think we could use as a common base for encodings. We'd `impl FromAddress for orchard::PaymentAddress`, for example (only defining the Orchard payment address conversion, for example), which would erase network-specific information where we don't need it, but the `zebra-chain` address type could retain it.

Reply:
(The above crate is focused on addresses rather than keys, but we could maybe generalise it for keys in a similar fashion. I'm a bit on-the-fence about whether I want key encodings in an address-focused crate, but I also don't want to see everything become a mish-mash of tiny crates...)